### PR TITLE
Unit details

### PR DESCRIPTION
**Pull request by huwshimi**
Added unit details to the inspector.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- view `http://0.0.0.0:6543/:flags:/react/?deploy-target=cs%3Atrusty%2Fwordpress`
- click wordpress
- click Units
- click on wordpress/0
- click Remove
- click Cancel
- click 'wordpress/0' in the inspector header to take you back to the unit list

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2944/
Test PASSed.

File: jujugui/static/gui/src/app/components/unit-details/_unit-details.scss [hunk 1]
```diff
@@ -0,0 +1,14 @@
+.unit-details {
+    &__properties {
+        padding: 10px 20px;
+
+        p {
+            @extend %stagger-fade-in;
+            margin: 0 0 10px 0;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/unit-details/_unit-details.scss]: Could this be converted to `margin-bottom: 10px` to stop overriding wanted styles later

File: jujugui/static/gui/src/app/components/unit-details/_unit-details.scss [hunk 2]
```diff
@@ -0,0 +1,14 @@
+.unit-details {
+    &__properties {
+        padding: 10px 20px;
+
+        p {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/unit-details/_unit-details.scss]: Can we target a class here please. All component styles should not target raw elements.

File: jujugui/static/gui/src/app/components/unit-list/test-unit-list.js [hunk 3]
```diff
@@ -49,10 +49,12 @@ describe('UnitList', () => {
       <juju.components.UnitListItem
         key={units[0].displayName}
         label={units[0].displayName}
+        action={output.props.children[1].props.children[1].props.action}
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/unit-list/test-unit-list.js]: This seems a little specific. Is there no way to target the props by class?

**Comment by anthonydillon** [comment]
Code and QA ok with some small comments on Sass. Luca and Spenser are doing a quick style review.

**Comment by Lucapaulina** [comment]
* text in the inspector should be #333
* There shouldn't be a confirmation dialogue on remove unit
* The line height for the destroy confirmation should be 24px
* Select all units row should have the row hover
* The cursor should change to the hand when you hover over the checkboxes
* The box shadow for the inspector blocks should be: 0 1px 4.85px .15px rgba(0, 0, 0, 0.2);
* The hover and the canvas is #f5f5f5 and it should be #f7f7f7
* The line height for the unit info should be 24px.
* The select all units row should be selectable across the whole unit.
* Clicking on the checkbox accesses unit info
* Can you remove the blue outline that gets added to everything you click?
* On the original service view for wordpress "Off" on the Expose row is slightly lower than the Expose label.

You should use Ant's prototype as a styling guide: http://demo.haus/~ant/inspector-animations/

File: jujugui/static/gui/src/app/components/unit-list/test-unit-list.js (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-list/test-unit-list.js]: This is a little trick I found to be able to write test assertions using a component output as the comparison. The issue we found was that these functions are internally bound to so there is no way to pass in a function which matches what this should be. Instead we reach deep into the React object and grab the function we expect to be there and use it to compare against.

**Comment by hatched** [comment]
@Lucapaulina that blue outline is a default style in Chrome on OSX thing. It is possible to remove it but it's also what indicates what's selected for 'tabbing'. Maybe its colour can be changed instead of removing entirely.

**Comment by Lucapaulina** [comment]
Could the hover state just activate if you tab through? I can see the
benefit of it for tabbing but it really sucks for general use. Can we
remove it for the time being and we'll make sure we explore a solution for
tabbing post-ODS.

File: jujugui/static/gui/src/app/components/inspector/test-inspector.js [hunk 4]
```diff
@@ -73,7 +75,34 @@ describe('Inspector', function() {
     var output = shallowRenderer.getRenderOutput();
     assert.deepEqual(output.props.children[1].props.children,
         <juju.components.UnitList
-          units={['units']}/>);
+          serviceId="demo"
+          units={['units']}
+          changeState={undefined} />);
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector/test-inspector.js]: we should pass a stub in on line 72 and then check for it here just to make sure it is getting passed through

File: jujugui/static/gui/src/app/components/unit-details/test-unit-details.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  });
+
+  it('renders the remove button', function() {
+    var output = jsTestUtils.shallowRender(
+      <juju.components.UnitDetails
+        unit={fakeUnit} />);
+    var buttons = [{
+      title: 'Remove',
+      action: output.props.children[1].props.buttons[0].action
+      }];
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-details/test-unit-details.js]: indentation -2

File: jujugui/static/gui/src/app/components/unit-details/test-unit-details.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+      <juju.components.ButtonRow
+        buttons={buttons} />);
+  });
+
+  it('renders the remove confirmation', function() {
+    var output = jsTestUtils.shallowRender(
+      <juju.components.UnitDetails
+        unit={fakeUnit} />);
+    var buttons = [
+        {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-details/test-unit-details.js]: indentations

File: jujugui/static/gui/src/app/components/unit-details/test-unit-details.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        'This cannot be undone.';
+    var shallowRenderer = jsTestUtils.shallowRender(
+      <juju.components.UnitDetails
+        unit={fakeUnit} />, true);
+    var output = shallowRenderer.getRenderOutput();
+      var buttons = [
+        {
+          title: 'Cancel',
+          action: output.props.children[2].props.buttons[0].action
+          },
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-details/test-unit-details.js]: indentations

File: jujugui/static/gui/src/app/components/unit-details/unit-details.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    _hideConfirmation: function() {
+      this.setState({confirmationOpen: false});
+    },
+
+    render: function() {
+      var unit = this.props.unit;
+      var buttons = [{
+        title: 'Remove',
+        action: this._showConfirmation
+        }];
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-details/unit-details.js]: indentations

File: jujugui/static/gui/src/app/components/unit-list-item/test-unit-list-item.js [hunk 9]
```diff
@@ -44,6 +44,43 @@ describe('UnitListItem', () => {
         <label htmlFor="unit-name-unit">unit-name</label>]);
   });
 
+  it('does not have a for id if it is a nav element', () => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-list-item/test-unit-list-item.js]: What? :)

File: jujugui/static/gui/src/app/components/unit-list/unit-list.js [hunk 10]
```diff
@@ -55,6 +55,28 @@ YUI.add('unit-list', function() {
...
 
     /**
+      The callable to be passed to the unit items for navigating to the unit
+      details.
+
+      @method _unitItemAction
+      @param {Object} e The click event.
+    */
+    _unitItemAction: function(e) {
+      var unitId = e.currentTarget.getElementsByTagName(
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-list/unit-list.js]: the id should probably be put as a data attribute, that way the whole thing doesn't fail if we decide to change how the unit names are displayed.

**Comment by huwshimi** [comment]
This is going to need another pretty thorough review as I've made a lot of changes.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2952/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2953/
Test PASSed.

File: jujugui/static/gui/src/app/assets/css/base.scss [hunk 11]
```diff
@@ -51,20 +51,20 @@ $font-family: 'Ubuntu', 'Helvetica Neue', Helvetica, Arial, sans-serif;
 $font-family-mono: 'Ubuntu Mono', 'Ubuntu-Mono', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace;
 
 /* Define common colors. */
-$canvas-background: #f5f5f5;
+$light-grey: #f7f7f7;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/assets/css/base.scss]: We already have a `$light-grey` var in base.scss

**Comment by Lucapaulina** [comment]
+1 LGTM

Awesome work Huw!

**Comment by anthonydillon** [comment]
Code and QA good :+1: thank you @huwshimi

File: jujugui/static/gui/src/app/assets/css/_animations.scss [hunk 12]
```diff
@@ -7,3 +7,23 @@
...
+    animation-name: fadeIn;
+    animation-fill-mode: forwards;
+    opacity: 0;
+}
+
+%stagger-fade-in {
+    @extend %fade-in;
+
+    @for $i from 1 through 99 {
+        $fade-in-stagger-speed: .01s;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/assets/css/_animations.scss]: The `$fade-in-stagger-speed` in the prototype is .03s

Comment (anthonydillon) [inline jujugui/static/gui/src/app/assets/css/_animations.scss]: As seen here: https://github.com/anthonydillon/inspector-animations/blob/master/static/scss/_states.scss#L133

File: jujugui/static/gui/src/test/test_ui_state.js [hunk 13]
```diff
@@ -1131,6 +1131,7 @@ describe('UI State object', function() {
     }];
 
     it('can generate proper urls from the state object', function() {
+      window.flags = {};
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_ui_state.js]: I'm not sure this is a good idea - this means that as soon as we remove the react flag these tests will fail and we won't necessarily remember why.

File: jujugui/static/gui/src/app/components/unit-list-item/unit-list-item.js [hunk 14]
```diff
@@ -53,6 +77,16 @@ YUI.add('unit-list-item', function() {
       this.setState({checked: checked});
     },
 
+    /**
+      Don't bubble the click event to the parent.
+
+      @method _stopBubble
+      @param {Object} The click event from the checkbox.
+    */
+    _stopBubble: function(e) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/unit-list-item/unit-list-item.js]: nice - thanks for adding a test for this as well.

File: jujugui/static/gui/src/app/assets/css/_animations.scss (hunk 12 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/assets/css/_animations.scss]: Oh, sorry I must have copied this line: https://github.com/anthonydillon/inspector-animations/blob/master/static/scss/_states.scss#L120

File: jujugui/static/gui/src/test/test_ui_state.js (hunk 13 above)
Comment (huwshimi) [inline jujugui/static/gui/src/test/test_ui_state.js]: Ah yeah. I'll update this to just remove the react flag.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2973/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks everyone! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**